test/test_init_teardown.c

```diff
@@ -0,0 +1,436 @@
+
+err_dlclose:
+    dlclose(h_umf);
+
+    return -1;
+}
+
+static void umf_unload(void) {
+    fprintf(stderr, "umf_unload: finalizing UMF ...\n");
+    umf_low_lat = NULL;
+    umf_high_bw = NULL;
+    umf_large_cap = NULL;
+    umf_default = NULL;
+
+    fprintf(stderr, "umf_unload: destroying umf memory pools ...\n");
+    if (umf_pool_low_lat) {
+        fprintf(stderr, "umf_unload: destroying the lowest latency umf "
+                        "memory pool ...\n");
+        (*umfPoolDestroy)(umf_pool_low_lat);
+        fprintf(stderr, "umf_unload: the lowest latency umf memory "
+                        "pool destroyed\n");
+    }
+
+    if (umf_pool_high_bw) {
+        fprintf(stderr, "umf_unload: destroying the highest bandwidth umf "
+                        "memory pool ...\n");
+        (*umfPoolDestroy)(umf_pool_high_bw);
+        fprintf(stderr, "umf_unload: the highest bandwidth umf "
+                        "memory pool destroyed\n");
+    }
+
+    if (umf_pool_large_cap) {
+        fprintf(stderr, "umf_unload: destroying the highest capacity umf "
+                        "memory pool ...\n");
+        (*umfPoolDestroy)(umf_pool_large_cap);
+        fprintf(stderr, "umf_unload: the highest capacity umf "
+                        "memory pool destroyed\n");
+    }
+
+    if (umf_pool_default) {
+        fprintf(stderr,
+                "umf_unload: destroying the default umf memory pool ...\n");
+        (*umfPoolDestroy)(umf_pool_default);
+        fprintf(stderr, "umf_unload: the default umf memory pool "
+                        "destroyed\n");
+    }
+
+    fprintf(stderr, "umf_unload: destroying umf memory providers ...\n");
+    if (umf_provider_low_lat) {
+        (*umfMemoryProviderDestroy)(umf_provider_low_lat);
+        fprintf(stderr, "umf_unload: the lowest latency umf memory "
+                        "provider destroyed\n");
+    }
+
+    if (umf_provider_high_bw) {
+        (*umfMemoryProviderDestroy)(umf_provider_high_bw);
+        fprintf(stderr, "umf_unload: the highest bandwidth umf "
+                        "memory provider destroyed\n");
+    }
+
+    if (umf_provider_large_cap) {
+        (*umfMemoryProviderDestroy)(umf_provider_large_cap);
+        fprintf(stderr, "umf_unload: the highest capacity umf "
+                        "memory provider destroyed\n");
+    }
+
+    if (umf_provider_default) {
+        (*umfMemoryProviderDestroy)(umf_provider_default);
+        fprintf(stderr, "umf_unload: the default umf memory "
+                        "provider destroyed\n");
+    }
+
+    // deinitialize libumf (decrement the reference counter of users)
+    fprintf(stderr, "umf_unload: calling umfTearDown() ...\n");
+    (*umfTearDown)();
+
+    fprintf(stderr, "umf_unload: closing umf library ...\n");
+    dlclose(h_umf);
+    fprintf(stderr, "umf_unload: umf library closed\n");
+}
+
+static int run_test(int wrong_dtor_order) {
+
+    fprintf(stderr, "\n\n*** Running test: wrong_dtor_order = %i ***\n\n\n",
```

Comment:
This could be an if-else with messages clearly stating if the order should be correct or not in the current run.

Reply:
Done.